AlertingNG: pause/unpause definitions via the API

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
It introduces two new endpoints for pausing and unpausing an alert definition:
`/api/alert-definitions/pause`
`/api/alert-definitions/unpause`
 
Paused alert definitions aren't scheduled to be evaluated on certain intervals but they can still be evaluated using the API.

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

Pause example:
```
curl -H "Content-Type: application/json" -XPOST http://admin:admin@localhost:3000/api/alert-definitions/pause -d '{
    "uids": ["mh_hJLfGz","unknown"]
}'
{"message":"1 alert definitions paused"}
```

Unpause example:
```
curl -H "Content-Type: application/json" -XPOST http://admin:admin@localhost:3000/api/alert-definitions/unpause -d '{
    "uids": ["mh_hJLfGz","unknown"]
}'
{"message":"1 alert definitions unpaused"}
```

papagian committed:
AlertingNG: pause/unpause definitions via the API

kylebrandt commented:
Looks good to me and works. This is scope creep, but I'm wondering if we should change this to accept a list of UIDs. This way, the API would be setup for pausing multiple alert definitions in one call.

Use cases:

- Future UI: Allow for multiple alerts in the list to be selected (e.g. checkboxes) and pause/unpause them all in one call.
- Administration automation. call the list api endpoint, filter with JQ, send single call with list to pause.

wbrowne commented on pkg/services/ngalert/api.go:
```diff
@@ -180,3 +182,33 @@ func (ng *AlertNG) unpauseScheduler() response.Response {
 	return response.JSON(200, util.DynMap{"message": "alert definition scheduler unpaused"})
 }
+
+// alertDefinitionPauseEndpoint handles PUT /api/alert-definitions/pause/:alertDefinitionUID.
```
```suggestion
// alertDefinitionPauseEndpoint handles POST /api/alert-definitions/pause/:alertDefinitionUID.
```

wbrowne commented on pkg/services/ngalert/api.go:
```diff
@@ -180,3 +182,33 @@ func (ng *AlertNG) unpauseScheduler() response.Response {
+	return response.JSON(200, util.DynMap{"message": "alert definition paused"})
+}
+
+// alertDefinitionUnpauseEndpoint handles PUT /api/alert-definitions/unpause/:alertDefinitionUID.
```
```suggestion
// alertDefinitionUnpauseEndpoint handles POST /api/alert-definitions/unpause/:alertDefinitionUID.
```

wbrowne reviewed (approved):
Tested and it looks good! 😄

papagian committed:
Apply suggestions from code review

Co-authored-by: Will Browne <wbrowne@users.noreply.github.com>

papagian committed:
Enable pausing/unpausing multiple definitions

papagian committed:
fixup

papagian commented:
> Looks good to me and works. This is scope creep, but I'm wondering if we should change this to accept a list of UIDs. This way, the API would be setup for pausing multiple alert definitions in one call.

I have changed the API calls to expect multiple definition UIDs and I have updated the PR description accordingly.

kylebrandt reviewed (approved):
> > Looks good to me and works. This is scope creep, but I'm wondering if we should change this to accept a list of UIDs. This way, the API would be setup for pausing multiple alert definitions in one call.
> 
> I have changed the API calls to expect multiple definition UIDs and I have updated the PR description accordingly.

Works for me. I was thinking could have the list in the url, but maybe url size issues etc, so I think I like this more.

papagian commented:
Closing and re-opening in hope to get the license/cla check green.
